src/test/unit/utils/async.test.ts

```diff
@@ -0,0 +1,59 @@
+import { strict as assert } from "assert";
+
+import { WaitGroup } from "../../../utils/async";
+
+function withTimeout<T>(promise: Promise<T>, timeout: number): Promise<T> {
+    return new Promise((resolve, reject) => {
+        setTimeout(() => {
+            reject(new Error("timed out"));
+        }, timeout);
+
+        promise.then(resolve);
+    });
+}
+
+describe("WaitGroup", () => {
+    it("should resolve immediately on a new object", async () => {
+        const wg = new WaitGroup();
+
+        await withTimeout(wg.promise, 100);
+    });
+
+    it("should wait forever if there are an imbalanced number of add calls", async () => {
```

Comment:
Semaphores have the same problem (one of many). Seems like this is reinventing semaphores, but nowhere does it mention that.

Reply:
https://www.npmjs.com/package/@jpwilliams/waitgroup has no dependencies and has some more docs + references to Golang analogs, so +1